CommandPalette: Search for dashboards using API

joshhunt opened the pull request:
**What is this feature?**

Adds Recent Dashboards and full search results to the command pallette:

 - AppWrapper.tsx will now decide whether to render on login page or not
 - Refactored actions into `useActions` hook to make the CommandPalette code a bit easier to read
 - Created a new `CommandPaletteAction` type to enforce priority and section are set on actions.
 - 5 recently viewed dashboards are shown at the top of the root Command Palette
 - Use search API to dynamically search for dashboards and add them to the list
 - Remove icons from Pages nav tree
 
| ![image](https://user-images.githubusercontent.com/46142/211550696-979b093b-8968-4557-b71f-cb8065f4f655.png) |  ![image](https://user-images.githubusercontent.com/46142/211550921-d0b48853-4cb1-4850-b228-73ef38a3a671.png) |
| ------|--------|


**Why do we need this feature?**

So all dashboards can be searched for via Command Palette

**Who is this feature for?**

People who search for dashboards

**Special notes for your reviewer**:

Fixes: #61092

Some product questions I have:

 - At the moment I remove the Starred and Recent dashboards when we have dashboard search results. Thoughts?
 - I'm not totally in love with the 'Dashboards...' name for the parent action. Thoughts on how to present these instead?

joshhunt committed:
CommandPalette: Search for dashboards using API

joshhunt committed:
Fix ordering of dashboards

zsofiakm commented:
> At the moment I remove the Starred and Recent dashboards when we have dashboard search results. Thoughts?
I think that's fair. We can display starred and recent dashboards when you open the command palette, before initiating a search. When you haven't found what you need among recent and starred items, you'll perform a search, so we no longer need to show those.
For the future, we should consider showing your recently viewed dashboards right when you open the command palette, without having to go into "Dashboards...". Ideally we should display the last 5-10 recently viewed pages, whether those were dashboards or other pages.

> I'm not totally in love with the 'Dashboards...' name for the parent action. Thoughts on how to present these instead?
I don't quite get the intention of this action. As far as I see, it's supposed to:
- Show recent and starred dashboards: Do we need this under an action or could we show them when the command palette is launched?
- You can perform a search filtered down to dashboards only. Not sure how much needed that is at the moment, since we're only searching dashboards and pages, and to perform an advanced search on dashboards, you should go to the dashboards page.

So the label would depend on the intention of this action, but I'm not sure if we even need it.

ashharrison90 reviewed (approved):
first off, this looks great! 🥳 

my 2c (i think it's similar to zsofia's):
- i'm not sure how much value we get from having a separate `Dashboards...` section. i'd be in favour of showing a cut-down `Recent` list at the top of the main list (e.g. recent 5? 10? dashboards at the top) and removing the starred section altogether - it already appears as a child of pages
- think we could show actions at the top level as well? section order for me would be something like:
  - Recent Dashboards
  - Contextual actions (see e.g. Explore)
  - Actions
  - Pages
  - Preferences

could absolutely merge this now as-is and make any changes in a follow-up PR btw 👌

joshhunt committed:
Put recent + search dashboards in root list, refactor actions into hook

joshhunt committed:
limit recent dashboards to 5

ashharrison90 commented on public/app/features/commandPalette/actions/useActions.ts:
```diff
@@ -0,0 +1,49 @@
+import { getDashboardSearchResultActions, getRecentDashboardActions } from './dashboardActions';
+import getStaticActions from './staticActions';
+
+const debouncedDashboardSearch = debounce(getDashboardSearchResultActions, 100);
```
think we should be a little more aggressive with our debouncing here btw, 100 is very short. 200 seems like a better compromise between frontend/backend perf? 🤔

joshhunt committed:
search debounce to 200ms

joshhunt commented:
Made a few changes from feedback:

 - Removed the nested 'Dashboards...' view
   - 5 Recent dashboards are showed at the top of the root view
   - Search results are also shown in the root (when searching)
 - Removed icons from Pages nav tree actions
 - 
and a bunch of refactors that came along for the ride

joshhunt committed:
update priorities

ashharrison90 commented on public/app/features/commandPalette/actions/staticActions.ts:
```diff
@@ -31,15 +25,15 @@ function navTreeToActions(navTree: NavModelItem[], parent?: NavModelItem): Actio
     }
 
-    const action: Action = {
+    const section = isCreateAction ? t('command-palette.section.actions', 'actions') : getPagesSectionTranslation();
```
think the default here should be `Actions`, also do you need to run i18n-extract as well?

joshhunt replied:
😭

ashharrison90 commented on public/app/features/commandPalette/values.ts:
```diff
@@ -0,0 +1,4 @@
+export const RECENT_DASHBOARDS_PRORITY = 4;
+export const DEFAULT_PRIORITY = 3;
+export const PREFERENCES_PRIORITY = 2;
+export const SEARCH_RESULTS_PRORITY = 1; // Dynamic actions should be below static ones so the list doesn't 'jump' when they come in
```
this comment doesn't seem to be correct. if i search for `new`, i see `Dashboards` then `actions` then `Recently viewed dashboards` 
![image](https://user-images.githubusercontent.com/20999846/211557521-bdae96d6-8dd6-495a-aa4d-ceb970d132cc.png)

don't feel super strongly about this right now as i suspect it's gonna change following our mob session/outcome of https://github.com/grafana/grafana/issues/61093 anyway

joshhunt replied:
Ahhh, you're right. I think the Section needs priority instead, which impacts the types i made, so I'll tackle this seperately.

ashharrison90 reviewed (approved):
lgtm, minor nit about translations

ashharrison90 commented:
also think we should only show `Recently viewed dashboards` at the top level only when no search string is present (if this is possible). again, doesn't have to hold out this PR. would prefer we get this in so we can build off it 🚀

joshhunt committed:
extract i18n
